orphan process: fix multi-staged process leak

connectdotz opened the pull request:
fixes #550 

This PR address the first scenario mentioned [here](https://github.com/jest-community/vscode-jest/issues/550#issuecomment-615930298):

> if I close the workspace before jest can complete the run, I see all jest child process became orphan processes ...

## Summary
Turns out if the first run-all-test process is stopped, it will not prevent the subsequent "watch" process from starting. This means if we close the workspace or stop the jest-process before the "watch" run starts, these watch processes will be leaked...

The fix here is to add an `isCancelled` check before starting the "watch" process and wrap the stop function to set this flag accordingly.

## TODO
To fully address #550, we also need jest-community/jest-editor-support#39. Once `jest-editor-support` released the new version, I will update the package.json here accordingly.

stephtr commented:
I have tried it on Windows and it seems like process management is still working fine.

I have to admit that this was my first detailed look at the Jest process management. One thing I noticed is that, at least to me, it would feel more natural if cancellation handling would be part of `JestProcess`, with an additional argument of the `exitCallback` describing the exit reason (whether `stop` was called).

connectdotz commented:
@stephtr I am so glad that you asked this question 👍 Originally I didn't think anybody other than `JestProcessManager` cares about how the process terminated so no point to spill that state to other places... but I was wrong, this state is also used by `JestExt` thus already in `JestProcess` (stopRequested()) and we don't even need to modify the callback signature.

The code is updated and I reverted some of the refactoring from `JestProcessManager` to minimize code change. Unfortunately had to update quite a bit tests to mock the `JestProcess` due to the new change. 

While debugging this thing I can't stand the `JestExt` reporting error without its identity, so added a one-line code change that is not related to this issue but should be pretty helpful for the multi-root projects...

also updated the `jest-editor-support` version to pick up the other orphan-process fix.

I am going to run this new build for a couple of days to make sure there is no unexpected issue and would appreciate some more reviews and real-world testing so we can ensure to rollout this critical bugfix ASAP...

connectdotz commented on src/JestProcessManagement/JestProcessManager.ts:
```diff
@@ -31,13 +31,15 @@ export class JestProcessManager {
-        })
-        exitCallback(exitedJestProcess, jestProcessInWatchMode)
+        if (!exitedJestProcess.stopRequested()) {
+          this.removeJestProcessReference(exitedJestProcess)
```
yeah decided to leave refactoring out, there are other JestProcess related refactoring in our technical debt list, could tackle them at a separate PR

stephtr replied:
In the meantime I would nevertheless move the `removeJestProcessReference` line out of the `if` block. Or am I missing something?

connectdotz replied:
If a process is stopped it would have been removed, see the stop()/stopAll() in `JestProcessManager`, therefore, only need to call `removeJestProcessReference` if it is not stopped.

stephtr replied:
Ah thanks, I didn't think of that.

connectdotz commented on src/JestExt.ts:
```diff
@@ -137,7 +137,9 @@ export class JestExt {
           if (!jestProcess.stopRequested()) {
-            let msg = `Starting Jest in Watch mode failed too many times and has been stopped.`
+            let msg = `(${
+              this.workspaceFolder.name
```
that's fair

connectdotz commented:
> On Windows it seems like everything is still working fine.

Were you able to reproduce the orphan process with the previous version when closing the workspace before the watch-mode can start? This bug should occur on windows as well...

stephtr commented:
> Were you able to reproduce the orphan process (scenario 1) with the previous version when closing the workspace before the watch-mode can start?

Yes, I was able to reproduce it.

seanpoulter commented on src/JestProcessManagement/JestProcessManager.ts:
```diff
@@ -31,13 +31,15 @@ export class JestProcessManager {
+            exitCallback,
+          })
+          exitCallback(exitedJestProcess, jestProcessInWatchMode)
+        }
```
**Non-blocking / Style**:
I'd prefer if we swapped out the `if (!` with a guard clause to explicitly communicate what happens when `stopRequested` is truthy. It's a bit more verbose, but it helps communicate what we're trying to do.
```js
        // When the user input caused the process to stop, stop.
        if (exitedJestProcess.stopRequested()) {
          return;
        }
```

seanpoulter commented on tests/JestProcessManagement/JestProcessManager.test.ts:
```diff
@@ -266,6 +264,56 @@ describe('JestProcessManager', () => {
+
+        eventEmitter.emit('debuggerProcessExit', p)
+        expect(jestProcessManager.numberOfProcesses).toBe(0)
+        expect(jestProcessMock.mock.instances.length).toBe(1)
```
It looks like you're asserting that a mock has been called `n` times. How do you feel about using one of the fancy matchers, `.toHaveBeenCalledTimes(number)` or the alias `.toBeCalledTimes`? Here are the [docs](https://jestjs.io/docs/en/expect#tohavebeencalledtimesnumber).

connectdotz commented:
@seanpoulter both are good points and easy to change so they are updated. 

Thanks for the review, merging now...
